### ISSUE TITLE
Fakes that overrides Equals are not considered to be equal to themselves 

### ISSUE THREAD

**Issue by afifit** [FakeItEasy/FakeItEasy#1774]
Similar to #1630, I found an odd behavior and couldn't find any info on it - 

```c#    
class MyClass {
    public override bool Equals(object o) { ... } // calls to base
    public override int GetHashCode() { ... } // calls to base
}
var myClass = A.Fake<MyClass>();
myClass.Should().Be(myClass); // this fails.
```
This seems odd to me as they are the same object. Can someone explain if it is the intended behavior or is it a bug?
I know that `CallsBaseMethods()` can solve this problem, but it is annoying to call it on every fake I create.

**Comment by blairconrad** [comment]
Hey, @afifit. Thanks for writing, and sorry you're having problems.

The explanation is that FakeItEasy is applying the same logic it would to any virtual method - it sees a method that returns a `bool` and you get `false` out of it. Before #1630 was fixed, it would've made no difference at all whether you had overridden the `Equals` . The change for that issue tries to identify the 3 `object` methods `Equals`, `GetHashCode`, and `ToString` and see if the intercepted method is one of those, and if so, it tries to be "nicer" and provide some better behaviour, like making the same objects be Equals.
However, the detection checks to see if they are exactly the same method, including that they are declared on the same class. So `MyClass.Equals` does not look like `Object.Equals`, and the special logic doesn't kick in. That's why you get the default "return false" behaviour.

We'll need to think a bit about whether intercepting overridden object methods is the right thing to do.

In the meantime, as you say, using `CallsBaseMethods` will work for you, but is annoying. Have you considered using [Implicit Creation Options](https://fakeiteasy.readthedocs.io/en/stable/implicit-creation-options/) to configure every `MyClass` that you fake, to avoid adding a lot of boilerplate? It might make a workaround more palatable.

**Comment by blairconrad** [comment]
Oh, and probably you've thought of this, but maybe just stop overriding methods solely to call the base method…

**Comment by afifit** [comment]
Thank you for the detailed answer, I think the creation options can help me, so I will try to use that for now. 
> Oh, and probably you've thought of this, but maybe just stop overriding methods solely to call the base method…
Well It was just an example - in my case there is a real value\reference equality checks in Equals.

**Comment by blairconrad** [comment]
>> The explanation is that FakeItEasy is applying the same logic it would to any virtual method - it sees a method that returns a bool and you get false out of it.

> Wait, then why not overriding Equals is working right? Object.Equals is a virtual method, and it doesn't seems like FakeItEasy returns false when calling it.

I should've written "it would to any virtual method besides non-overridden `Object` methods". My next paragraph was intended to explain the exception made for overridden `Object` methods, but I did not make it clear in the intro. Sorry.

> Maybe overriding Equals is a special case, but it really feels wrong when an instance doesn't equal to itself.

I agree that it's surprising, but my private personal opinion, which may not align with my public-as-one-of-the-faces-of-FakeItEasy opinion, is that Fakes are kind of different objects, and we have a set of reasonable rules built in for how intercepted methods are handled (generally, return a default value, although that's a gross simplification), and that the more special cases we add, the more confusing it gets. Granted, `Object` methods are pretty fundamental, so maybe they deserve them, but then we open up additional complexities, like how to deal with overridden methods, methods that wrap, etc. Sometimes I think that an initial surprise ("`Equals` returned false") plus a very simple explanation ("everything returns false") is preferable to behaviour that's a little more intuitive most of the time but is nearly impossible to keep in your head once you're out of the 90% case.

I have other opinions around how to write tests using Fakes, and they often end up with me saying things like "there shouldn't be much reason to ever compare the Fakes anyhow", but everyone's use cases are different, so I try not to spout those too much in public!

**Comment by thomaslevesque** [comment]
Hi @afifit,

If you don't override `Equals`, FakeItEasy provides a default behavior that makes sense, as @blairconrad mentioned. However, if you *do* override it, it implies there's a special logic for deciding if two instances are equal. FakeItEasy doesn't know that logic, so it couldn't provide a reasonable behavior that would do what you want. Using `CallsBaseMethod` is the correct approach in this situation.

So, I don't think this is a bug, or something that we should change.

**Comment by thomaslevesque** [comment]
Hi @afifit,

@blairconrad and I discussed this further, and we came to the conclusion that the current behavior probably isn't useful to anyone, so we're going to change it (for `Equals`, but also for `GetHashCode` and `ToString`). I opened a new issue to track this (#1777).

Note that the fake still won't have the overridden behavior (to get that behavior, use `CallsBaseMethod` as you've been doing), but at least a fake will be equal to itself.
